### PR TITLE
Initial PR for backport of AzureARM from Salt 2018.3 to OpenSUSE-Salt 2016.11.4

### PR DESCRIPTION
**Pull request by edlane**
## NOTICE: This PR should be considered preliminary...
While it has been tested WRT AzureARM salt-cloud functionality it should be given more scrutiny and widespread Salt test coverage before acceptance.

### What does this PR do?
provides initial backport of Salt's AzureARM cloud provider from 2018.3 to 2016.11.4

### What issues does this PR fix or reference?
see https://bugzilla.suse.com/show_bug.cgi?id=1087342

### Previous Behavior
AzureARM was not supported in 2016.11.4

### New Behavior
AzureARM IS supported in 2016.11.4

### Tests written?
Yes -- see Testcase referenced in attachments here:
https://bugzilla.suse.com/show_bug.cgi?id=1087342

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

### CODE REVIEW

**Comment by mcalmer** [comment]
We must test if this code break python 2.6 compatibility.  The 2016.11.4 package need to run ol SLE11.
It is not required that this feature run there, but the minion must not crash on startup when the lazy loader iterate over all modules and call virtual.

**Comment by kkaempf** [comment]
@isbm @brejoc @dincamihai @meaksh please review, merge, and run through the testsuite.

File: salt/utils/stringutils.py [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+    # This shouldn't be six.string_types because if we're on PY2 and we already
+    # have a string, we should just return it.
+    if isinstance(s, str):
+        return s
+    if six.PY3:
+        if isinstance(s, (bytes, bytearray)):
+            # https://docs.python.org/3/howto/unicode.html#the-unicode-type
+            # replace error with U+FFFD, REPLACEMENT CHARACTER
+            return s.decode(encoding or __salt_system_encoding__, "replace")
+        raise TypeError('expected str, bytes, or bytearray not {}'.format(type(s)))
```
Comment (isbm) [inline salt/utils/stringutils.py]: This is not Python 2.6 compatible code.

Comment (rjschwei) [reply salt/utils/stringutils.py]: @isbm why do you say the code at this point is not compatible with Python 2.6?

salt provides six and the code in question, which if unguarded would be python 2.6 incompatible, is guarded by a

if six.PY3:

There is a Python 2.6 incompatibility, but that is in the is_binary() function on line 161 where:

trans_args = ({ord(x): None for x in text_characters},)

is certainly a syntactical problem fro Python 2.6.

@edlane please merge https://github.com/edlane/salt-1/pull/1 into you branch and update this PR ASAP! we need to continue with review and testing.

**Comment by kkaempf** [comment]
@brejoc (@dincamihai @meaksh) please review and merge ! (@isbm is on vacation :unamused: )

**Comment by brejoc** [comment]
SR to products:testing can be found here: https://build.opensuse.org/request/show/597337

**Comment by brejoc** [comment]
@edlane The SR got accepted this afternoon. You can find the update RPMs in [products:testing](https://build.opensuse.org/package/show/systemsmanagement:saltstack:products:testing/salt). Happy testing!

**Comment by kkaempf** [comment]
Thanks @brejoc, highly appreciated !

**Pull request merged**